pkg/nar/dump.go

```diff
@@ -0,0 +1,123 @@
+package nar
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"syscall"
+)
+
+// DumpPath will serialize a path on the local file system to NAR format,
+// and write it to the passed writer.
+func DumpPath(w io.Writer, path string) error {
+	// initialize the nar writer
+	nw, err := NewWriter(w)
+	if err != nil {
+		return err
+	}
+
+	// make sure the NAR writer is always closed, so the underlying goroutine is stopped
+	defer nw.Close()
+
+	err = dumpPath(nw, path, "")
+	if err != nil {
+		return err
+	}
+
+	return nw.Close()
+}
+
+// dumpPath recursively calls itself for every node in the path.
+func dumpPath(nw *Writer, path string, subpath string) error {
+	// assemble the full path.
+	p := filepath.Join(path, subpath)
+
+	// peek at the path
+	fi, err := os.Lstat(p)
+	if err != nil {
+		return err
+	}
+
+	if fi.Mode()&os.ModeSymlink == os.ModeSymlink {
+		linkTarget, err := os.Readlink(p)
+		if err != nil {
+			return err
+		}
+
+		// write the symlink node
+		err = nw.WriteHeader(&Header{
+			Path:       subpath,
+			Type:       TypeSymlink,
+			LinkTarget: linkTarget,
+		})
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	if fi.IsDir() {
+		// write directory node
+		err := nw.WriteHeader(&Header{
+			Path: subpath,
+			Type: TypeDirectory,
+		})
+		if err != nil {
+			return err
+		}
+
+		// look at the children
+		files, err := os.ReadDir(filepath.Join(path, subpath))
+		if err != nil {
+			return err
+		}
+
+		// loop over all elements
+		for _, file := range files {
+			err := dumpPath(nw, path, filepath.Join(subpath, file.Name()))
+			if err != nil {
+				return err
+			}
+		}
+
+		return nil
+	}
+
+	if fi.Mode().IsRegular() {
+		// write regular node
+		err := nw.WriteHeader(&Header{
+			Path: subpath,
+			Type: TypeRegular,
+			Size: fi.Size(),
+			// If it's executable by the user, it'll become executable.
+			// This matches nix's dump() function behaviour.
+			Executable: fi.Mode()&syscall.S_IXUSR != 0,
+		})
+		if err != nil {
+			return err
+		}
+
+		// open the file
+		f, err := os.Open(p)
```

Comment:
Maybe we should consider a `defer f.Close()` to prevent FD exhaustion here: we might end up opening a bunch of files in a single GC interval when dumping a large tree.

Reply:
:+1: